### PR TITLE
RUBY-1934 Auto encryption and decryption unit tests

### PR DESCRIPTION
**Pull request by egiurleo**
This PR implements the methods necessary for auto encryption/decryption and tests those methods.

Upcoming PRs will actually integrate those methods into the required database commands.

### CODE REVIEW

File: lib/mongo/crypt/encryption_io.rb [hunk 1]
```diff
@@ -51,7 +54,37 @@ def find_keys(filter)
...
+      # Get collection info for a collection matching the provided filter
+      #
+      # @param [ Hash ] filter
+      #
+      # @return [ Hash ] The collection information
+      def collection_info(filter)
+        result = @client.database.list_collections
+
+        name = filter['name']
+        result.find { |r| r['name'] == name }
```
Comment (egiurleo) [inline lib/mongo/crypt/encryption_io.rb]: @p-mongo I was trying to do something like Python's list_collections(filter) method -- https://api.mongodb.com/python/current/api/pymongo/database.html#pymongo.database.Database.list_collections. What do you think is the best approach?

File: lib/mongo/crypt/context.rb [hunk 2]
```diff
@@ -106,6 +114,11 @@ def raise_from_status
         status.raise_crypt_error
       end
 
+      # Indicate that state machine is done feeding I/O responses back to libmongocrypt
+      def mongo_done
```
Comment (p-mongo) [inline lib/mongo/crypt/context.rb]: `mongo_xxx` method names to me imply that they are interacting with the server. If they are interacting with mongocryptd/libmongocryptd can the prefix be changed to `mongocrypt`, `lmc`, `mc`, etc.?

File: lib/mongo/crypt/encryption_io.rb (hunk 1 above)
Comment (p-mongo) [reply lib/mongo/crypt/encryption_io.rb]: The server command supports a filter argument (https://docs.mongodb.com/manual/reference/command/listCollections/). Our helper doesn't apparently (https://api.mongodb.com/ruby/current/Mongo/Database.html#collections-instance_method). Amending the helper to pass the filter seems like a valid strategy, but if the inline filtering here works it seems fine to do this as well.

File: lib/mongo/crypt/auto_encrypter.rb [hunk 3]
```diff
@@ -71,9 +73,52 @@ def setup_encrypter(options = {})
...
                                 @encryption_options[:mongocryptd_uri],
                                 monitoring_io: mongocryptd_client_monitoring_io,
+                                server_selection_timeout: mongocryptd_server_selection_timeout, # For testing purposes
                               )
 
-        # TODO: use all the other options for auto-encryption/auto-decryption
+        # Tests fail with live background threads if the @mongocryptd client is not closed at the
+        # end of every test run. Better to write one line registering it as a local client than to
+        # do it separately in every test.
+        ClientRegistry.instance.register_local_client(@mongocryptd_client) if defined?(ClientRegistry)
```
Comment (p-mongo) [inline lib/mongo/crypt/auto_encrypter.rb]: Can you please make a ticket to investigate why this is needed and preferably remove it before release?

Comment (egiurleo) [reply lib/mongo/crypt/auto_encrypter.rb]: https://jira.mongodb.org/browse/RUBY-2069

**Pull request merged**